README.md
inwx.com XML-RPC PHP Client
=========
You can access all functions of our frontend via an application programming interface (API). Our API is based on the XML-RPC protocol and thus can be easily addressed by almost all programming languages. The documentation and programming examples in PHP, Java, Ruby and Python can be downloaded here.

There is also an OT&E test system, which you can access via ote.inwx.com. Here you will find the known web interface which is using a test database. On the OT&E system no actions will be charged. So you can test how to register domains etc.

Documentation
------
You can view a detailed description of the API functions in our documentation. The documentation as PDF ist part of the Projekt. You also can read the documentation online http://www.inwx.de/en/help/apidoc

Installation
-------
We provide composer support for an easy installation. If u wish to setup our php-client manually
you can also require the domrobot class without composer.

```
composer require inwx/domrobot
```

API Endpoints
-------
Production endpoint: `https://api.domrobot.com/xmlrpc/`

Development endpoint: `https://api.ote.domrobot.com/xmlrpc/`

Example
-------

```php
require "domrobot.class.php";

// Config
$endpoint = "https://api.ote.domrobot.com/xmlrpc/";
$username = "your_username";
$password = "your_password";

// Create Domrobot instance and authorize
$domrobot = new INWX\Domrobot($endpoint);
$domrobot->setDebug(false);
$domrobot->setLanguage('en');
$result = $domrobot->login($username, $password);

// Build an API Call
if ($result['code'] == 1000) {
	$result = $domrobot->call('domain', 'check', [
	    'domain' => 'mydomain.com'
	]);
	
	echo '<pre>';
	print_r($result);
	
} else {
    echo '<pre>';
	print_r($res);
}

// Logout
$result = $domrobot->logout();
```

You can also take a look at the example.php in this project.

License
----

MIT


Changelog.md
#XML-RPC Inwx-Domrobot 
 
XML-RPC support in PHP is not enabled by default. 
You will need to use the --with-xmlrpc[=DIR] configuration option when compiling PHP to enable XML-RPC support.
 
#Changelog:

##2019/02/02 - Master
- Add composer setup documentation
- Update API docbook

##2015/01/XX
 - Add Composer
 - Move Domrobot Class to Namespace "INWX"
 - Set Cookiefile default to the System tmp folder/file
 - Add Funktion setCookiefile and getCookiefile

##2014/10/29
 - Use Namespace and add Composer

##2013/01/22 - v2.3
 - added google 2-step-verification methods
 - added parameter 'sharedSecret' to login method
 
##2012/04/08 - v2.2
 - use "nested" methods  (e.g. domain.check)
 - removed nonce and secure-login
 - added setter and getter
 - added credentials params to login function 
 - response utf-8 decoded 
 - removed newlines and white spaces in xml request (verbosity=no_white_space)
 - added optional clTRID set/get functions

##2011/07/19 - v2.1 
 - using cookiefile instead of session
 - added login and logout function
 - added client version transmission
    
   
by InterNetworX Ltd. & Co. KG


example.php
<?php
/*
 * XML-RPC support in PHP is not enabled by default. 
 * You will need to use the --with-xmlrpc[=DIR] configuration option when compiling PHP to enable XML-RPC support. 
 * This extension is bundled into PHP as of 4.1.0.
 * 
 * cURL needs to be installed and activated
 * 
 */

header('Content-type: text/plain; charset=utf-8');
error_reporting(E_ALL);

require "INWX/Domrobot.php";

//$addr = "https://api.domrobot.com/xmlrpc/";
//$addr = "https://api.ote.domrobot.com/xmlrpc/";

// Alternatively, you may use JSON (experimental):
//
//$addr = "https://api.domrobot.com/jsonrpc/";
//$addr = "https://api.ote.domrobot.com/jsonrpc/";

// Remeber when using JSON you get objects back, not arrays, so use:
//   $res->code
// instead of:
//   $res['code']

$usr = "your_username";
$pwd = "your_password";

$domrobot = new INWX\Domrobot($addr);
$domrobot->setDebug(false);
$domrobot->setLanguage('en');
$res = $domrobot->login($usr,$pwd);

if ($res['code'] == 1000) {
	$obj = "domain";
	$meth = "check";
	$params = [];
	$params['domain'] = "mydomain.com";
	$res = $domrobot->call($obj,$meth,$params);
	print_r($res);
} else {
	print_r($res);
}

$res = $domrobot->logout();

INWX/Domrobot.php
<?php

namespace INWX;

class Domrobot
{
    private $debug = false;
    private $address;
    private $language;
    private $customer = false;
    private $clTRID = null;

    private $_ver = "2.4.1";
    private $_cookiefile = NULL;

    function __construct($address)
    {
        $this->address = (substr($address, -1) != "/") ? $address . "/" : $address;
        $this->_cookiefile = tempnam(sys_get_temp_dir(), 'INWX');
    }

    public function setLanguage($language)
    {
        $this->language = $language;
    }

    public function getLanguage()
    {
        return $this->language;
    }

    public function setDebug($debug = false)
    {
        $this->debug = (bool) $debug;
    }

    public function getDebug()
    {
        return $this->debug;
    }

    public function setCookiefile($file)
    {
        if ((file_exists($file) && !is_writable($file)) || (!file_exists($file) && !is_writeable(dirname($file)))) {
            throw new \Exception("Cannot write cookie file: '" . $file . "'. Please check file/folder permissions.", 2400);
        }

        $this->_cookiefile = $file;
    }

    public function getCookiefile()
    {
        return $this->_cookiefile;
    }

    public function setCustomer($customer)
    {
        $this->customer = (string)$customer;
    }

    public function getCustomer()
    {
        return $this->customer;
    }

    public function setClTrId($clTrId)
    {
        $this->clTRID = (string)$clTrId;
    }

    public function getClTrId()
    {
        return $this->clTRID;
    }

    public function login($username, $password, $sharedSecret = null)
    {
        $fp = fopen($this->_cookiefile, "w");
        fclose($fp);

        if (!empty($this->language)) {
            $params['lang'] = $this->language;
        }
        $params['user'] = $username;
        $params['pass'] = $password;

        $loginRes = $this->call('account', 'login', $params);
        if (!empty($sharedSecret) && $loginRes['code'] == 1000 && !empty($loginRes['resData']['tfa'])) {
            $_tan = $this->_getSecretCode($sharedSecret);
            $unlockRes = $this->call('account', 'unlock', ['tan' => $_tan]);
            if ($unlockRes['code'] == 1000) {
                return $loginRes;
            } else {
                return $unlockRes;
            }
        } else {
            return $loginRes;
        }
    }

    public function logout()
    {
        $ret = $this->call('account', 'logout');
        if (file_exists($this->_cookiefile)) {
            unlink($this->_cookiefile);
        }
        return $ret;
    }

    public function call($object, $method, array $params = [])
    {
        if (isset($this->customer) && $this->customer != "") {
            $params['subuser'] = $this->customer;
        }
        if (!empty($this->clTRID)) {
            $params['clTRID'] = $this->clTRID;
        }

        $useJSON = preg_match('!/jsonrpc!', $this->address);

        if ($useJSON) {
            $request = json_encode(['method' => $object . "." . $method, 'params' => $params]);
        } else {
            $request = xmlrpc_encode_request(strtolower($object . "." . $method), $params, ["encoding" => "UTF-8", "escaping" => "markup", "verbosity" => "no_white_space"]);
        }


        $header[] = "Content-Type: text/xml";
        $header[] = "Connection: keep-alive";
        $header[] = "Keep-Alive: 300";
        $header[] = "X-FORWARDED-FOR: " . @$_SERVER['HTTP_X_FORWARDED_FOR'];
        $ch = curl_init();
        curl_setopt($ch, CURLOPT_URL, $this->address);
        curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false);
        curl_setopt($ch, CURLOPT_TIMEOUT, 65);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_HTTPHEADER, $header);
        curl_setopt($ch, CURLOPT_COOKIEFILE, $this->_cookiefile);
        curl_setopt($ch, CURLOPT_COOKIEJAR, $this->_cookiefile);
        curl_setopt($ch, CURLOPT_POST, true);
        curl_setopt($ch, CURLOPT_POSTFIELDS, $request);
        curl_setopt($ch, CURLOPT_USERAGENT, "DomRobot/{$this->_ver} (PHP " . phpversion() . ")");

        $response = curl_exec($ch);
        curl_close($ch);
        if ($this->debug) {
            echo "Request:\n" . $request . "\n";
            echo "Response:\n" . $response . "\n";
        }

        if ($useJSON) {
            return json_decode($response);
        } else {
            return xmlrpc_decode($response, 'UTF-8');
        }
    }

    private function _getSecretCode($secret)
    {
        $_timeSlice = floor(time() / 30);
        $_codeLength = 6;

        $secretKey = $this->_base32Decode($secret);
        // Pack time into binary string
        $time = chr(0) . chr(0) . chr(0) . chr(0) . pack('N*', $_timeSlice);
        // Hash it with users secret key
        $hm = hash_hmac('SHA1', $time, $secretKey, true);
        // Use last nipple of result as index/offset
        $offset = ord(substr($hm, -1)) & 0x0F;
        // grab 4 bytes of the result
        $hashPart = substr($hm, $offset, 4);

        // Unpak binary value
        $value = unpack('N', $hashPart);
        $value = $value[1];
        // Only 32 bits
        $value = $value & 0x7FFFFFFF;

        $modulo = pow(10, $_codeLength);
        return str_pad($value % $modulo, $_codeLength, '0', STR_PAD_LEFT);
    }

    private function _base32Decode($secret)
    {
        if (empty($secret)) return '';

        $base32chars = $this->_getBase32LookupTable();
        $base32charsFlipped = array_flip($base32chars);

        $paddingCharCount = substr_count($secret, $base32chars[32]);
        $allowedValues = [6, 4, 3, 1, 0];
        if (!in_array($paddingCharCount, $allowedValues)) return false;
        for ($i = 0; $i < 4; $i++) {
            if ($paddingCharCount == $allowedValues[$i] &&
                substr($secret, -($allowedValues[$i])) != str_repeat($base32chars[32], $allowedValues[$i])) return false;
        }
        $secret = str_replace('=', '', $secret);
        $secret = str_split($secret);
        $binaryString = "";
        for ($i = 0; $i < count($secret); $i = $i + 8) {
            $x = "";
            if (!in_array($secret[$i], $base32chars)) return false;
            for ($j = 0; $j < 8; $j++) {
                $x .= str_pad(base_convert(@$base32charsFlipped[@$secret[$i + $j]], 10, 2), 5, '0', STR_PAD_LEFT);
            }
            $eightBits = str_split($x, 8);
            for ($z = 0; $z < count($eightBits); $z++) {
                $binaryString .= (($y = chr(base_convert($eightBits[$z], 2, 10))) || ord($y) == 48) ? $y : "";
            }
        }
        return $binaryString;
    }

    private function _getBase32LookupTable()
    {
        return [
            'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', // 7
            'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', // 15
            'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', // 23
            'Y', 'Z', '2', '3', '4', '5', '6', '7', // 31
            '=' // padding char
        ];
    }
}

domainlist_successively.php
<?php
/*
 * XML-RPC support in PHP is not enabled by default.
 * You will need to use the --with-xmlrpc[=DIR] configuration option when compiling PHP to enable XML-RPC support.
 * This extension is bundled into PHP as of 4.1.0.
 *
 * cURL needs to be installed and activated
 *
 */

header('Content-type: text/plain; charset=utf-8');
error_reporting(E_ALL);
require "INWX/Domrobot.php";


$addr = "https://api.domrobot.com/xmlrpc/";
//$addr = "https://api.ote.domrobot.com/xmlrpc/";

$usr = "";
$pwd = "";

$domains_per_request = 20;
$sleep_between_requests = false; // sleep for a defined number of seconds after each api-request (default: false)


$domrobot = new INWX\Domrobot($addr);
$domrobot->setDebug(false);
$domrobot->setLanguage('en');
$res = $res_lgn = $domrobot->login($usr,$pwd);

$domains = array();

if($res_lgn['code'] == 1000){
	$res_count = $domrobot->call("domain", "list", array("page" => 1, "pagelimit" => 1));
	$num_domains = (int) $res_count['resData']['count'];
	$pages = round(($num_domains / $domains_per_request)+2, 0);
	for($i=1;$i<$pages;$i++){
		$res = $domrobot->call("domain", "list", array("page" => $i, "pagelimit" => $domains_per_request));
		foreach($res['resData']['domain'] as $domain){
			$domains[] = $domain;
		}
		if($sleep_between_requests !== false && is_numeric($sleep_between_requests)) sleep((int) $sleep_between_requests);
	}
	print_r($domains);
} else{
	print_r($res_lgn);
}

domrobot.class.php
<?php
/*
 * XML-RPC Inwx-Domrobot 
 * 
 * XML-RPC support in PHP is not enabled by default. 
 * You will need to use the --with-xmlrpc[=DIR] configuration option when compiling PHP to enable XML-RPC support.
 * 
 * Changelog:
 * ----------
 * 
 * 2013/01/22 - v2.3
 * 	  - added google 2-step-verification methods
 * 	  - added parameter 'sharedSecret' to login method
 * 
 * 2012/04/08 - v2.2
 *    - use "nested" methods  (e.g. domain.check)
 *    - removed nonce and secure-login
 *    - added setter and getter
 *    - added credentials params to login function 
 *    - response utf-8 decoded 
 * 	  - removed newlines and white spaces in xml request (verbosity=no_white_space)
 *    - added optional clTRID set/get functions
 * 
 * 2011/07/19 - v2.1 
 * 	  - using cookiefile instead of session
 * 	  - added login and logout function
 * 	  - added client version transmission
 *    
 *   
 * by InterNetworX Ltd. & Co. KG
 */

class domrobot 
{
	private $debug=false;
	private $address;
	private $language;
	private $customer=false;
	private $clTRID = null;

	private $_ver = "2.3";
	private $_cookiefile = "domrobot.tmp";

	function __construct($address) {
		$this->address = (substr($address,-1)!="/")?$address."/":$address;

		$seperator = (DIRECTORY_SEPARATOR=="/" || DIRECTORY_SEPARATOR=="\\")?DIRECTORY_SEPARATOR:"/";
		$this->_cookiefile = dirname(__FILE__).$seperator.$this->_cookiefile;

		if (file_exists($this->_cookiefile) && !is_writable($this->_cookiefile) ||
			!file_exists($this->_cookiefile) && !is_writeable(dirname(__FILE__))) {
			throw new Exception("Cannot write cookiefile: '{$this->_cookiefile}'. Please check file/folder permissions.",2400);			
		}
	}

	public function setLanguage($language) {
		$this->language = $language;
	}
	public function getLanguage() {
		return $this->language;
	}
	
	public function setDebug($debug=false) {
		$this->debug = (bool)$debug;
	}
	public function getDebug() {
		return $debug;
	}
	
	public function setCustomer($customer) {
		$this->customer = (string)$customer;
	}
	public function getCustomer() {
		return $this->customer;
	}
	
	public function setClTrId($clTrId) {
		$this->clTRID = (string)$clTrId;
	}
	public function getClTrId() {
		return $this->clTRID;
	}

	public function login($username,$password,$sharedSecret=null) {
        $fp = fopen($this->_cookiefile, "w");
        fclose($fp);
        
        if (!empty($this->language)) {
			$params['lang'] = $this->language;
        }
		$params['user'] = $username;
		$params['pass'] = $password;
		
		$loginRes = $this->call('account','login',$params); 
		if (!empty($sharedSecret) && $loginRes['code']==1000 && !empty($loginRes['resData']['tfa'])) {
			$_tan = $this->_getSecretCode($sharedSecret);
			$unlockRes = $this->call('account','unlock',array('tan'=>$_tan));
			if ($unlockRes['code']==1000) {
				return $loginRes;
			} else {
				return $unlockRes;
			}
		} else {
			return $loginRes;
		}
	}

	public function logout() {
		$ret = $this->call('account','logout');
		if (file_exists($this->_cookiefile)) {
			unlink($this->_cookiefile);
		}
		return $ret;
	}

	public function call($object, $method, array $params=array()) {
		if (isset($this->customer) && $this->customer!="") {
			$params['subuser'] = $this->customer;
		}
		if (!empty($this->clTRID)) {
			$params['clTRID'] = $this->clTRID;
		}
		
		$request = xmlrpc_encode_request(strtolower($object.".".$method), $params, array("encoding"=>"UTF-8","escaping"=>"markup","verbosity"=>"no_white_space"));
	
		$header[] = "Content-Type: text/xml";   
		$header[] = "Connection: keep-alive";
		$header[] = "Keep-Alive: 300";
		$header[] = "X-FORWARDED-FOR: ".@$_SERVER['HTTP_X_FORWARDED_FOR'];
		$ch = curl_init();
		curl_setopt($ch,CURLOPT_URL,$this->address);
		curl_setopt($ch,CURLOPT_SSL_VERIFYPEER, false); 
		curl_setopt($ch,CURLOPT_TIMEOUT,65);
        curl_setopt($ch,CURLOPT_RETURNTRANSFER,true);
		curl_setopt($ch,CURLOPT_HTTPHEADER,$header);
		curl_setopt($ch,CURLOPT_COOKIEFILE,$this->_cookiefile);
		curl_setopt($ch,CURLOPT_COOKIEJAR,$this->_cookiefile);
		curl_setopt($ch,CURLOPT_POST,true);
        curl_setopt($ch,CURLOPT_POSTFIELDS,$request);
		curl_setopt($ch,CURLOPT_USERAGENT,"DomRobot/{$this->_ver} (PHP ".phpversion().")");

		$response = curl_exec($ch);
		curl_close($ch); 
		if ($this->debug) {
			echo "Request:\n".$request."\n";
			echo "Response:\n".$response."\n";
		}

		return xmlrpc_decode($response,'UTF-8');
	}
	
	private function _getSecretCode($secret) {
		$_timeSlice = floor(time() / 30);
		$_codeLength = 6;
	
		$secretKey = $this->_base32Decode($secret);
		// Pack time into binary string
		$time = chr(0).chr(0).chr(0).chr(0).pack('N*', $_timeSlice);
		// Hash it with users secret key
		$hm = hash_hmac('SHA1', $time, $secretKey, true);
		// Use last nipple of result as index/offset
		$offset = ord(substr($hm, -1)) & 0x0F;
		// grab 4 bytes of the result
		$hashPart = substr($hm, $offset, 4);
	
		// Unpak binary value
		$value = unpack('N', $hashPart);
		$value = $value[1];
		// Only 32 bits
		$value = $value & 0x7FFFFFFF;
	
		$modulo = pow(10, $_codeLength);
		return str_pad($value % $modulo, $_codeLength, '0', STR_PAD_LEFT);
	}
	
	private function _base32Decode($secret) {
		if (empty($secret)) return '';
	
		$base32chars = $this->_getBase32LookupTable();
		$base32charsFlipped = array_flip($base32chars);
	
		$paddingCharCount = substr_count($secret, $base32chars[32]);
		$allowedValues = array(6, 4, 3, 1, 0);
		if (!in_array($paddingCharCount, $allowedValues)) return false;
		for ($i = 0; $i < 4; $i++){
			if ($paddingCharCount == $allowedValues[$i] &&
					substr($secret, -($allowedValues[$i])) != str_repeat($base32chars[32], $allowedValues[$i])) return false;
		}
		$secret = str_replace('=','', $secret);
		$secret = str_split($secret);
		$binaryString = "";
		for ($i = 0; $i < count($secret); $i = $i+8) {
			$x = "";
			if (!in_array($secret[$i], $base32chars)) return false;
			for ($j = 0; $j < 8; $j++) {
				$x .= str_pad(base_convert(@$base32charsFlipped[@$secret[$i + $j]], 10, 2), 5, '0', STR_PAD_LEFT);
			}
			$eightBits = str_split($x, 8);
			for ($z = 0; $z < count($eightBits); $z++) {
				$binaryString .= ( ($y = chr(base_convert($eightBits[$z], 2, 10))) || ord($y) == 48 ) ? $y:"";
			}
		}
		return $binaryString;
	}
	
	private function _getBase32LookupTable() {
		return array(
				'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', // 7
				'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', // 15
				'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', // 23
				'Y', 'Z', '2', '3', '4', '5', '6', '7', // 31
				'=' // padding char
		);
	}
}
?>